packages/chrono-core/src/processors/processor.ts

```diff
@@ -11,38 +11,44 @@ export type TaskErrorEvent = {
   timestamp: Date;
 };
 
-/**
- * Emits 'error' event when a task exits because of an error.
- */
-export class TaskErrorChannel extends EventEmitter {
-  emitError(taskIndex: number, error: Error): void {
-    this.emit('error', { taskIndex, error, timestamp: new Date() } as TaskErrorEvent);
-  }
-
-  onError(handler: (errorEvent: TaskErrorEvent) => void): void {
-    this.on('error', (event: TaskErrorEvent) => handler(event));
-  }
-}
-
 /**
  * This class is a wrapper to its task which emits an error event if the task fails.
  */
-export class TaskRunner {
+export class TaskRunner extends EventEmitter {
   private taskIndex: number;
   private task: () => Promise<void>;
-  private errorChannel: TaskErrorChannel;
 
-  constructor(taskIndex: number, errorChannel: TaskErrorChannel, task: () => Promise<void>) {
+  constructor(taskIndex: number, task: () => Promise<void>) {
+    super();
+
     this.taskIndex = taskIndex;
     this.task = task;
-    this.errorChannel = errorChannel;
   }
 
   async run(): Promise<void> {
     try {
       await this.task();
     } catch (error) {
-      this.errorChannel.emitError(this.taskIndex, error as Error);
+      const errorEvent: TaskErrorEvent = {
+        taskIndex: this.taskIndex,
+        error: error as Error,
+        timestamp: new Date(),
+      };
+
+      this.emit('error', errorEvent);
+
+      // To prevent emitting `exit` event
+      return;
     }
+
+    this.emit('exit', this.taskIndex);
+  }
+
+  onceExit(): Promise<void> {
+    return new Promise((resolve) => this.once('exit', resolve));
+  }
+
+  onError(handler: (errorEvent: TaskErrorEvent) => void): void {
```

Comment:
so i like these methods alot. What i dont like is making the entire TaskRunner an emitter. but i cant put why into words